### PR TITLE
refactor(ext/node): add NodeEnv::Fs associated type

### PR DESCRIPTION
**Pull request by bartlomieju**
This commit adds associated type to "NodeEnv" trait, called "Fs".

The "Fs" type has a trait bound on "NodeFs", which specifies APIs required
for all ops and resolution APIs to function.

A "RealFs" implementation of "NodeFs" is exported from the "deno_node"
crate, that provides a default implementation for the trait.

All code in "deno_node" extension was changed to use the "NodeFs" trait
to handle file system operations, instead of relying on APIs from the standard
library.

### CODE REVIEW

File: ext/node/lib.rs [hunk 1]
```diff
@@ -41,18 +42,33 @@ pub use resolution::DEFAULT_CONDITIONS;
...
 
 pub trait NodePermissions {
   fn check_read(&mut self, path: &Path) -> Result<(), AnyError>;
 }
 
-// TODO(bartlomieju):
-// pub trait NodeFs {
-//   fn current_dir() -> Result<PathBuf, AnyError>;
-// }
+pub trait NodeFs {
```
Comment (dsherret) [inline ext/node/lib.rs:52]: Seeing this used now, this means that the implementation of this trait needs to use global data/functions only, right? I think that should be ok, but it's a bit limiting.

File: ext/node/lib.rs
Comment (dsherret) [inline ext/node/lib.rs:1]: Can we add a clippy rule to this crate to ban certain functions like stuff under std::env?

File: ext/node/lib.rs (hunk 1 above)
Comment (bartlomieju) [reply ext/node/lib.rs:52]: Did it for the first pass - if we decide that having state is needed we can change it. Doesn't seem to be necessary though

File: ext/node/lib.rs
Comment (bartlomieju) [reply ext/node/lib.rs:1]: It's somewhat possible using: https://rust-lang.github.io/rust-clippy/master/#disallowed_method but I will have to list all methods from `std::fs` manually

Comment (bartlomieju) [reply ext/node/lib.rs:1]: Added `clippy.toml` that disallows these methods

**Review by dsherret — APPROVED** [review]
LGTM!

**Pull request merged**